File: Project4/julia/src/metamodel.jl
module Metamodel
export Metadata, writedata!

using JSON

mutable struct Metadata
    latticesize::T           where T<:Integer
    MCiterations::T          where T<:Integer
    temperature::T           where T<:Real
    saveperiod::T            where T<:Integer
    J::T                     where T<:Integer
    h::T                     where T<:Real
    initialorientation::Symbol
    basepath::String
    energypath::String
    magneticmomentpath::String
    evolutionpath::String
    metadatapath::String
    seed::T                  where T<:Real
    rng::MersenneTwister

    energydim::Tuple
    magneticmomentdim::Tuple
    evolutiondim::Tuple

    function Metadata{I<:Integer, R<:Real}(size::I, iter::I, temp::R, orientation::Symbol,
                      path::String; seed=rand(Int))
        # Construct a new model by checking that each argument is valid
        # and using default values for several variables

        size ≤ 0 && throw(ArgumentError("The lattice size must be positive."))
        iter ≤ 0 && throw(ArgumentError("The number of MC iterations must be positive."))
        temp ≤ 0 && throw(ArgumentError("The temperature [K] must be positive."))
        seed ≤ 0 && throw(ArgumentError("The random seed must be positive"))
        orientation ∉ [:up, :down, :random] && throw(ArgumentError("The orientation must be :up, :down or :random"))

        # All IO paths
        if !endswith(path, "/")
            path = path * "/"
        end
        energypath = "energies.bin"
        evolutpath = "evolution.bin"
        momentpath = "magneticmoment.bin"
        metadatapath = "meta.json"

        # Construct the object
        new(size, iter, temp, 1000, 1, 0, orientation,
            path, energypath, momentpath, evolutpath, metadatapath,
            seed, MersenneTwister(seed), (0,), (0,), (0,0))
    end
end

function savedimensions!(model::Metadata, evolution, energy, magneticmoment)
    model.energydim = size(energy)
    model.magneticmomentdim = size(magneticmoment)
    model.evolutiondim = size(evolution)
end

function writedata!(model::Metadata, evolution::Array{Int8, 3}, energy::Array{T}, magneticmoment::Array{T}) where T<:Real
    output = [(model.energypath, energy), (model.magneticmomentpath, magneticmoment)]
    for (path, data) in output
        open(x -> write(x, data), model.basepath * path, "w")
    end

    stream = open(model.basepath * model.evolutionpath, "w")
    for m in 1:size(evolution, 1)
        write(stream, evolution[m, :, :])
    end

    savedimensions!(model, evolution, energy, magneticmoment)
    writemetadata(model)
end

function writedata!(model::Metadata, evolution::Array{Int8, 2}, energy::T, magneticmoment::T) where T<:Real
    output = [(model.energypath, energy),
              (model.magneticmomentpath, magneticmoment),
              (model.evolutionpath, evolution)]
    for (path, data) in output
        stream = open(model.basepath * path, "w")
        write(stream, data)
    end
    savedimensions!(model, evolution, energy, magneticmoment)
    writemetadata(model)
end

function writemetadata(model::Metadata)
    stream = open(model.basepath * model.metadatapath, "w")
    magneticmeta = Dict("path" => model.magneticmomentpath,
                        "dim"  => model.magneticmomentdim,
                        "type" => "float64")
    energymeta = Dict("path" => model.energypath,
                      "dim"  => model.energydim,
                      "type" => "float64")
    evolutionmeta = Dict("path" => model.evolutionpath,
                         "dim"  => model.evolutiondim,
                         "type" => "int8")
    metadata = Dict("energy"          => energymeta,
                    "magnetic moment" => magneticmeta,
                    "evolution"       => evolutionmeta,
                    "saveperiod"      => model.saveperiod,
                    "seed"            => model.seed,
                    "lattice size"    => model.latticesize,
                    "MC iterations"   => model.MCiterations)
    JSON.print(stream, metadata)
end

end

File: Project4/cpp/src/metamodel.cpp
#include <json/json.h>
#include <json/writer.h>
#include <iostream>
#include <fstream>
#include <type_traits>
#include "metamodel.h"

void Metamodel::write() const{
  std::ofstream metafile(basepath+metapath);
  Json::Value root;
  root["evolution"]["dim"]  = Json::arrayValue;
  root["evolution"]["type"] = "int64";
  root["evolution"]["path"] = "evolution.bin";
  root["evolution"]["dim"].append(M);
  root["evolution"]["dim"].append(N);
  root["evolution"]["dim"].append(N);

  root["magnetic moment"]["dim"]  = Json::arrayValue;
  root["magnetic moment"]["type"] = "int32";
  root["magnetic moment"]["path"] = "magneticmoment.bin";
  root["magnetic moment"]["dim"].append(M);

  root["flips"]["dim"]  = Json::arrayValue;
  root["flips"]["type"] = "int32";
  root["flips"]["path"] = flippath;
  root["flips"]["dim"].append(M);

  root["energy"]["dim"]  = Json::arrayValue;
  root["energy"]["type"] = "float64";
  root["energy"]["path"] = "energies.bin";
  root["energy"]["dim"].append(M);

  root["saveperiod"]    = 1000;
  root["seed"]          = seed;
  root["lattice size"]  = N;
  root["MC iterations"] = M;
  root["parallel"]      = parallel;
  metafile << root << std::endl;
  metafile.close();
  //TODO: ENDRE SAVEPERIOD TIL NUMBER OF SAVES
}

void Metamodel::read(const std::string& filename) {
    std::ifstream parameters(filename);
    if (!parameters.good())
        throw std::runtime_error("Could not find parameters");
    Json::Value root;
    parameters >> root;
    Tstart             = root["Tstart"].asDouble();
    Tstop              = root["Tstop"].asDouble();
    Tnumber            = root["Tnumber"].asDouble();
    Tstep              = (Tstop - Tstart)/Tnumber;
    parallel           = root["parallel"].asBool();
    seed               = root["seed"].asInt();
    N                  = root["lattice size"].asInt();
    M                  = root["monte carlo cycles"].asInt();
    initialOrientation = root["initial orientation"].asString();
}

void Metamodel::save(std::vector<arma::imat>& states, std::vector<double>& energies,
                     std::vector<int>& magmoments, std::vector<int>& nFlips) const{
  // Write energy to file
  std::ofstream energyStream;
  energyStream.open(basepath + energypath, std::ios::out | std::ios::binary);
  binaryDump(energyStream, energies);
  energyStream.close();

  // Write magnetic moment to file
  std::ofstream magmom;
  magmom.open(basepath + magneticmomentpath, std::ios::out | std::ios::binary);
  binaryDump(magmom, magmoments);
  magmom.close();

  // Write system states to file
  std::ofstream evoStream;
  evoStream.open(basepath + evolutionpath, std::ios::out | std::ios::binary);
  binaryDump(evoStream, states);
  evoStream.close();

  // Write number of accepted flips each MC cycle to file
  std::ofstream flips;
  flips.open(basepath + flippath, std::ios::out | std::ios::binary);
  binaryDump(flips, nFlips);
  flips.close();
}

template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
void Metamodel::binaryDump(std::ofstream& stream, const std::vector<T>& container) const{
    stream.write((char*)&container[0], M*sizeof(container[0]));
}

template<typename T>
void Metamodel::binaryDump(std::ofstream& stream, const std::vector<arma::Mat<T>>& container) const{
    for(const auto& matrix: container){
        stream.write((char*)&matrix(0,0), N*N*sizeof(matrix(0,0)));
    }
}

void Metamodel::saveExpectationValues(std::ofstream& stream, std::vector<double>& expVals,
                                      double T, int numProcessors) const{
  // Dumps values for a given temperature to stream.
  unsigned int Mprime   = M*numProcessors;
  double       factor   = 1.0/static_cast<double>(Mprime);
  double       spinNorm = 1.0/static_cast<double>(N*N);

  double expectE        = expVals[0]*factor;
  double expectESquared = expVals[1]*factor;
  double expectM        = expVals[2]*factor;
  double expectMSquared = expVals[3]*factor;
  double expectMFabs    = expVals[4]*factor;
  double n_flips        = expVals[5]*factor;

  double varE = (expectESquared - expectE*expectE)*spinNorm;
  double varM = (expectMSquared - expectMFabs*expectMFabs)*spinNorm;

  double Cv  = varE/T/T;
  double sus = varM/T;

  stream << T << " "
         << expectE*spinNorm << " "
         << expectESquared*spinNorm << " "
         << varE << " "
         << Cv << " "
         << expectM*spinNorm <<  " "
         << expectMSquared*spinNorm << " "
         << expectMFabs*spinNorm << " "
         << varM << " "
         << sus << " "
         << n_flips << "\n";
}
